hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/state/Schema.java

```diff
@@ -0,0 +1,110 @@
+ * whenever a new {@link ReadableKVState} is to be created, or an existing one removed, or a
+ * migration has to happen. If your service makes use of a forwards and backwards compatible
+ * serialization system (such as protobuf), then it is not necessary to define a new {@link Schema}
+ * for simple compatible changes to serialization.
+ */
+public abstract class Schema implements Comparable<Schema> {
+    /** The version of this schema */
+    private final SoftwareVersion version;
+
+    /**
+     * Create a new instance
+     *
+     * @param version The version of this schema
+     */
+    protected Schema(@NonNull SoftwareVersion version) {
+        this.version = Objects.requireNonNull(version);
+    }
+
+    /**
+     * Gets the version of this {@link Schema}
+     *
+     * @return The version
+     */
+    @NonNull
+    public SoftwareVersion getVersion() {
+        return version;
+    }
+
+    /**
+     * Gets a {@link Set} of state definitions for states to create in this schema. For example,
+     * perhaps in this version of the schema, you need to create a new state FOO. The set will have
+     * a {@link StateDefinition} specifying the metadata for that state.
+     *
+     * <p>If a state is defined in this set that already exists, it will be redefined based on the
+     * returned values. This can be used if a state exists but needs to use a new {@link Serdes},
+     * for example.
+     *
+     * @return A map of all states to be created. Possibly empty.
+     */
+    @NonNull
+    @SuppressWarnings("rawtypes")
+    public abstract Set<StateDefinition> statesToCreate();
+
+    /**
+     * Called after all new states have been created (as per {@link #statesToCreate()}), this method
+     * is used to perform all <b>synchronous</b> migrations of state. This method will always be
+     * called with the {@link ReadableStates} of the previous version of the {@link Schema}. If
+     * there was no previous version, then {@code previousStates} will be empty, but not null.
+     *
+     * @param previousStates The {@link ReadableStates} of the previous {@link Schema} version
+     * @param newStates {@link WritableStates} for this schema.
+     */
+    public abstract void migrate(
+            @NonNull ReadableStates previousStates, @NonNull WritableStates newStates);
+
+    /**
+     * The {@link Set} of state keys of all states that should be removed <b>AFTER</b> {@link
+     * #migrate(ReadableStates, WritableStates)}.
+     *
+     * @return the set of states to remove
+     */
+    @NonNull
+    public abstract Set<String> statesToRemove();
+
+    /** {@inheritDoc */
+    @Override
+    public int compareTo(Schema o) {
+        return this.version.compareTo(o.version);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (!(o instanceof Schema other)) return false;
+
+        // BasicSoftwareVersion does not implement equals, so we will do a comparison to see if
```

Comment:
I have replaced `SoftwareVersion` with `SemanticVersion`, which is an application-level versioning schema. It is defined in the protobuf schema because we use it to return the semantic version through a HAPI query. So I will reuse it here. Ultimately we'll use the one created by PBJ, but for now I'll use the google generated one.